### PR TITLE
AugDiskCachedDataset to map the copy index to augmentation parameter

### PR DESCRIPTION
**Pull request by MinaKh**
This branch added a child class for `DiskCachedDataset` called `AugDiskCachedDataset`. 
Its main use is for a family of so-called deterministic augmentations with a rather discrete parameter space. For instance a noise augmentation on audio samples in which SNR can have only 5 values.

- In `DiskCachedDataset` num_copies can be used to generate N copies of a data sample. This is ok when used transforms/augmentations have an infinite/probabilistic parameter space. So the chance of generating repetitive augmented versions is very low. 
-  On the other hand for deterministic augmentations with N parameter it is advantageous to map the copy index to the parameter to avoid re-generating existing samples and to make sure generated copies cover all desired parameter space  . 
-  The main feature of this class is that the index of file copy is mapped to the parameter of augmentation

### CODE REVIEW

**Comment by biphasic** [comment]
Hello @MinaKh! Currently I don't understand how what you're trying to achieve with this class cannot already be done with existing classes. 
It seems to me that you want to control the augmentations exactly, but then I don't understand why they're called augmentations. Can you please 
* provide an example of how you use your proposed class
* explain with a concrete example why the current code cannot do what you need to do

Before I can merge this, this class would need a test as well, it might be helpful to add that as well.

**Comment by MinaKh** [comment]
> Hello @MinaKh! Currently I don't understand how what you're trying to achieve with this class cannot already be done with existing classes. It seems to me that you want to control the augmentations exactly, but then I don't understand why they're called augmentations. Can you please
> 
> * provide an example of how you use your proposed class
> * explain with a concrete example why the current code cannot do what you need to do
> 
> Before I can merge this, this class would need a test as well, it might be helpful to add that as well.

Hi @biphasic! Thanks for your feedback. 

- Deterministic augmentations are not uncommon, specially in audio processing and they are still called augmentation but in a more controlled way. 
- I have added a notebook to `docs/tutorails/Aug_DiskCachDataset.ipynb` and have addressed your raised point there with a synthetic dataset. Please let me know if it is not clear. 
- I also have added a test test/test_aug_caching.py which is pretty similar to what I presented in the notebook. Please let me know if you have other ideas for tests.

**Review by biphasic — CHANGES_REQUESTED** [review]
Hello Mina, I have two small changes that I request, then I can merge this

File: test/requirements.txt [hunk 1]
```diff
@@ -4,3 +4,4 @@ matplotlib
 hdf5plugin
 imageio
 aedat
+typing-extensions
```
[the remark this answers is not in the document]
Comment (MinaKh) [reply test/requirements.txt]: Thanks @biphasic I pushed updated branch based on your last comments.

File: test/test_aug_caching.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import os
+
+import numpy as np
+from torch.utils.data import Dataset
+from torchvision.transforms import Compose
```
Comment (biphasic) [inline test/test_aug_caching.py:5]: could you please also move this line

File: test/test_aug_caching.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import os
+
+import numpy as np
+from torch.utils.data import Dataset
```
Comment (biphasic) [inline test/test_aug_caching.py:4]: Do you need any functionality of this parent class? Could this be omitted?

Comment (MinaKh) [reply test/test_aug_caching.py]: if you mean Dataset: it was not necessary and I removed it. Thanks!

File: test/test_aug_caching.py (hunk 2 above)
Comment (MinaKh) [reply test/test_aug_caching.py]: done!

**Pull request merged**